lib/logstash/outputs/elasticsearch.rb

```diff
@@ -426,55 +536,53 @@ def generate_jks cert_path
     [jks.path, pwd]
   end
 
-  public
-  def receive(event)
-    return unless output?(event)
-
-    # Set the 'type' value for the index.
-    if @index_type
-      type = event.sprintf(@index_type)
-    else
-      type = event["type"] || "logs"
-    end
-
-    index = event.sprintf(@index)
-
-    document_id = @document_id ? event.sprintf(@document_id) : nil
-    buffer_receive([event.sprintf(@action), { :_id => document_id, :_index => index, :_type => type }, event.to_hash])
-  end # def receive
-
-  def flush(actions, teardown=false)
+  private
+  # in charge of submitting any actions in @retry_queue that need to be 
+  # retried
+  #
+  # This method is not called concurrently. It is only called by @retry_thread
+  # and once that thread is ended during the teardown process, a final call 
+  # to this method is done upon teardown in the main thread.
+  def retry_flush()
+    # it is only necessary to lock this mutex for the process of waiting for a 
+    # signal, and does not need to wrap the rest of the flush execution because 
+    # we ensure this method is to not be called concurrently.
+    @retry_flush_mutex.lock
     begin
-      @logger.debug? and @logger.debug "Sending bulk of actions to client[#{@client_idx}]: #{@host[@client_idx]}"
-      @current_client.bulk(actions)
-    rescue => e
-      @logger.error "Got error to send bulk of actions to elasticsearch server at #{@host[@client_idx]} : #{e.message}"
-      raise e
+      @retry_queue_needs_flushing.wait(@retry_flush_mutex) unless @retry_teardown_requested.true?
     ensure
-      unless @protocol == "node"
-          @logger.debug? and @logger.debug "Shifting current elasticsearch client"
-          shift_client
-      end
+      @retry_flush_mutex.unlock
     end
```

Comment:
maybe rewrite it like this?
```ruby
if @retry_teardown_requested.false?
    @retry_flush_mutex.lock
    begin
      @retry_queue_needs_flushing.wait(@retry_flush_mutex)
    ensure
      @retry_flush_mutex.unlock
    end
end
```

now, since this is not a tight loop, we could use `synchronize` instead of lock/unlock like this?

```ruby
@retry_flush_mutex.synchronize { @retry_queue_needs_flushing.wait(@retry_flush_mutex) } unless @retry_teardown_requested.true?
```